src/java.base/share/classes/java/io/SerializationMisdeclarationChecker.java

```diff
@@ -0,0 +1,300 @@
+ * questions.
+ */
+
+package java.io;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Arrays;
+
+import static jdk.internal.event.SerializationMisdeclarationEvent.*;
+import static java.lang.reflect.Modifier.*;
+
+final class SerializationMisdeclarationChecker {
+
+    private static final String SUID_NAME = "serialVersionUID";
+    private static final String SERIAL_PERSISTENT_FIELDS_NAME = "serialPersistentFields";
+    private static final String WRITE_OBJECT_NAME = "writeObject";
+    private static final String READ_OBJECT_NAME = "readObject";
+    private static final String READ_OBJECT_NO_DATA_NAME = "readObjectNoData";
+    private static final String WRITE_REPLACE_NAME = "writeReplace";
+    private static final String READ_RESOLVE_NAME = "readResolve";
+
+    private static final Class<?>[] WRITE_OBJECT_PARAM_TYPES = {ObjectOutputStream.class};
+    private static final Class<?>[] READ_OBJECT_PARAM_TYPES = {ObjectInputStream.class};
+    private static final Class<?>[] READ_OBJECT_NO_DATA_PARAM_TYPES = {};
+    private static final Class<?>[] WRITE_REPLACE_PARAM_TYPES = {};
+    private static final Class<?>[] READ_RESOLVE_PARAM_TYPES = {};
+
+    private final Class<?> cl;
+
+    SerializationMisdeclarationChecker(Class<?> cl) {
+        this.cl = cl;
+    }
+
+    @SuppressWarnings("removal")
+    void checkMisdeclarations() {
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            privilegedCheckSerialVersionUID();
+            privilegedCheckSerialPersistentFields();
+
+            privilegedCheckPrivateMethod(WRITE_OBJECT_NAME, WRITE_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(READ_OBJECT_NAME, READ_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(READ_OBJECT_NO_DATA_NAME, READ_OBJECT_NO_DATA_PARAM_TYPES, Void.TYPE);
+
+            privilegedCheckAccessibleMethod(WRITE_REPLACE_NAME, WRITE_REPLACE_PARAM_TYPES, Object.class);
+            privilegedCheckAccessibleMethod(READ_RESOLVE_NAME, READ_RESOLVE_PARAM_TYPES, Object.class);
+
+            return null;
+        });
+    }
+
+    private void privilegedCheckSerialVersionUID() {
+        Field f = declaredField(cl, SUID_NAME);
+        if (f == null) {
+            if (isOrdinaryClass()) {
+                commitEvent(SUID_EXPLICIT,
+                        "a private static final long " + SUID_NAME +
+                                " field should be declared");
+            }
+            return;
+        }
+        if (cl.isEnum()) {
+            commitEvent(SUID_INEFFECTIVE_ENUM,
+                    SUID_NAME + " is not effective in an enum class");
+        }
+        if (!isPrivate(f)) {
+            commitEvent(SUID_PRIVATE,
+                    SUID_NAME + " should be declared private");
```

Comment:
Rest of the event messages use "... to be effective", should this one too use a similar text?

Reply:
The field needs not be declared `private` to be effective, although it is recommended to do so. In this sense, "should" is just a recommendation, while "must" is really needed to make something effective.